web/src/engine/websites/MeianPlus.ts

```diff
@@ -0,0 +1,160 @@
+import { Tags } from '../Tags';
+import icon from './MeianPlus.webp';
+import { Chapter, DecoratableMangaScraper, type MangaPlugin, Manga, Page } from '../providers/MangaPlugin';
+import * as Common from './decorators/Common';
+import { Fetch, FetchWindowScript } from '../platform/FetchProvider';
+import type { Priority } from '../taskpool/DeferredTask';
+import DeScramble from '../transformers/ImageDescrambler';
+import { Exception } from '../Error';
+import { WebsiteResourceKey as R } from '../../i18n/ILocale';
+
+type APISeries = {
+    licences: APISerieDetails[];
+}
+
+type APISerie = {
+    licence: APISerieDetails
+}
+
+type APISerieDetails = {
+    id_licence: number,
+    titre_licence: string,
+    articles: {
+        ref: number,
+        titre: string,
+        ebook_statut: boolean
+    }[]
+}
+
+type APIEbook = {
+    success: boolean,
+    message: string,
+    ebook: {
+        max_page: number
+    }
+}
+
+type APIImages = {
+    images: {
+        ref: number,
+        key: string,
+        w: number,
+        h: number,
+        param: string
+    }[]
+}
+
+type BlockInfo = {
+    w: number,
+    h: number
+}
+
+const tokenScript = `decodeURIComponent(document.cookie.match(/token_meian_plus=([^;]+)/).at(1)).replaceAll('"', '');`;
+const scramblingMatrix: Array<Array<number>> = [];
+
+for (let i = 1; i <= 10; i++) {
+    for (let j = 1; j <= 10; j++) {
+        scramblingMatrix.push([i, j]);
+    }
+}
+
+export default class extends DecoratableMangaScraper {
+    private token: string = undefined;
+    private readonly apiUrl = 'https://api.meian-plus.fr/v1/';
+    private readonly imageCDN = 'https://ebook.meian-plus.fr/';
+
+    public constructor() {
+        super('meianplus', 'Meian Plus', 'https://www.meian-plus.fr', Tags.Media.Manga, Tags.Language.French, Tags.Source.Official);
+    }
+
+    public override get Icon() {
+        return icon;
+    }
+
+    private async UpdateToken() {
+        this.token = await FetchWindowScript(new Request(this.URI), tokenScript, 2500);
+    }
+
+    public override ValidateMangaURL(url: string): boolean {
+        return new RegExpSafe(`^${this.URI.origin}/catalogue/licence/[^/]+/\\d+$`).test(url);
+    }
+
+    public override async FetchManga(provider: MangaPlugin, url: string): Promise<Manga> {
+        const idLicence = url.match(/\/licence\/[^/]+\/(\d+)$/).at(1);
+        const { licence: { id_licence, titre_licence } } = await this.FetchJSON<APISerie>(this.CreateRequest(`./licence/?id_licence=${idLicence}`));
```

Comment:
Apply DRY principle, introduce a method to `FetchAPI` and create the request inside this method
Update `this.token` from within this method (only when necessary)
Don't use `this.token` outside the `FetchAPI` method

Reply:
> Update this.token from within this method (only when necessary)

should i use base url provides for checking? As its only necessary when gathering pages.